### PR TITLE
8333766: Stack overflow with anonymous class in super() parameter

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch fixes an issue in `LambdaToMethod` where the wrong code is generated to access a symbol in the enclosing scope.

`LambdaTranslationContext::addSymbol` has some logic to avoid referring to `T.this` when `T` is a type under construction. This fix works well when a lambda is directly enclosed in a `super` or `this` call, but does not scale to more complex cases, such as the one in this issue.

A more complete fix needs to check whether the target `T.this` is really available via a chain of enclosing `this$n` fields. If that's not the case, then `T.this` must be captured by the lambda more directly, as there's no *path* to it from the innermost enclosing this.

To detect that, I've added a method (`outerThisReachable`) which more or less mimic the lookup logic in `Lower::makeOwnerThisN`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333766](https://bugs.openjdk.org/browse/JDK-8333766): Stack overflow with anonymous class in super() parameter (**Bug** - P4)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19604/head:pull/19604` \
`$ git checkout pull/19604`

Update a local copy of the PR: \
`$ git checkout pull/19604` \
`$ git pull https://git.openjdk.org/jdk.git pull/19604/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19604`

View PR using the GUI difftool: \
`$ git pr show -t 19604`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19604.diff">https://git.openjdk.org/jdk/pull/19604.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19604#issuecomment-2155526431)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java [hunk 1]
```diff
@@ -2116,11 +2116,12 @@ public Symbol baseSymbol() {
             }
 
             void addSymbol(Symbol sym, LambdaSymbolKind skind) {
-                if (skind == CAPTURED_THIS && sym != null && sym.kind == TYP && !typesUnderConstruction.isEmpty()) {
+                if (skind == CAPTURED_THIS && sym != null && sym.kind == TYP) {
                     ClassSymbol currentClass = currentClass();
-                    if (currentClass != null && typesUnderConstruction.contains(currentClass)) {
-                        // reference must be to enclosing outer instance, mutate capture kind.
-                        Assert.check(sym != currentClass); // should have been caught right in Attr
+                    if (currentClass != null && currentClass != sym && !outerThisReachable(currentClass.type, sym)) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:2121]: There is a subtle change here - note that now if `currentClass == sym` we just skip this part (in this case just using `this` is enough).

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java [hunk 2]
```diff
@@ -2130,6 +2131,17 @@ void addSymbol(Symbol sym, LambdaSymbolKind skind) {
                 }
             }
 
+            private boolean outerThisReachable(Type current, Symbol target) {
+                if (current == Type.noType || typesUnderConstruction.contains(current.tsym)) {
+                    return false;
+                } else if (current.tsym.isSubClass(target, types)) {
+                    return true;
+                } else {
+                    return current.tsym.hasOuterInstance() &&
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:2140]: note that `hasOuterInstance` takes into account the `NOOUTERTHIS` flag, that is, whether the inner class can have an enclosing instance at all.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java [hunk 3]
```diff
@@ -2130,6 +2131,17 @@ void addSymbol(Symbol sym, LambdaSymbolKind skind) {
                 }
             }
 
+            private boolean outerThisReachable(Type current, Symbol target) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:2134]: The idea here is: if we can reach the target symbol from the current one w/o traversing any type under construction, or any type for which `NOOUTERTHIS` is set, then we can consider the target `T.this` reachable.

**Review by lahodaj — APPROVED** [review]
Makes sense to me.

**Comment by mcimadamore** [comment]
Closing this PR. This issue is related to other issues such as https://bugs.openjdk.org/browse/JDK-8334037, which can only be fixed reliably with a more invasive refactoring which moves Lower ahead of LambdaToMethod.

**Pull request closed without merging**